### ISSUE TITLE
Validation script fails for > 1 added packages

### ISSUE THREAD

**Issue by finestructure** [SwiftPackageIndex/SwiftPackageIndex-Server#847]

**Comment by finestructure** [comment]
Works fine when run locally, even without a token:

```

❯ xcrun swift validate.swift
Warning: Using anonymous authentication -- may run into rate limiting issues

+ https://github.com/guykogus/CodableGeoJSON.git
+ https://github.com/guykogus/CodableJSON.git

2 package(s) passed
```

**Comment by finestructure** [comment]
Online test passes as well: https://github.com/SwiftPackageIndex/PackageList/pull/792

My best guess is that validation fails when the nightly has exhausted the rate limit. That should also mess up runs with single additions, though.

Is it perhaps a red herring that we're observing it with multiple packages and they just happened to be submitted at times in the morning (UCT/CET) when the nightly was still in progress?

In any case, I'll look into porting over the rate limit wait/retry mechanism we're using in the nightly.

**Comment by finestructure** [comment]
Actually, the error message doesn't look like it's been a rate limit at all:

```
+ https://github.com/guykogus/CodableGeoJSON.git
+ https://github.com/guykogus/CodableJSON.git
🚨 https://github.com/guykogus/CodableJSON.git: dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: "The given data was not valid JSON.", underlyingError: Optional(Error Domain=NSCocoaErrorDomain Code=3840 "No value." UserInfo={NSDebugDescription=No value.})))
10
```

Might be a package specific problem.

**Comment by finestructure** [comment]
Cannot reproduce:

```
❯ xcrun swift validate.swift
Warning: Using anonymous authentication -- may run into rate limiting issues

+ https://github.com/guykogus/CodableGeoJSON.git
+ https://github.com/guykogus/CodableJSON.git

2 package(s) passed
```

At this point I think the best we can do is wait for another issue with an error and debug when it's happening. Specifically: check if the nightly is still running and making a `/rate_limit` call:

```yml
# rate-limit.restfile

variables:
  api_url: https://api.github.com

requests:
  repos:
    url: ${api_url}/rate_limit
    headers:
      Authorization: token ${GITHUB_TOKEN}
    validation:
      status: 200
    log:
      - json.rate
```

```
env GITHUB_TOKEN=... rester rate-limit.restfile
```

**Comment by daveverwer** [comment]
We're all sleeping while the nightly runs though. 😬

We could simulate it by kicking off a nightly run manually, and then when it's been processing long enough to get to the rate limit, submit a test package.

**Comment by Sherlouk** [comment]
Could we add an extra step to CI so when it fails (and only when it fails!) it runs a cURL command to download the current rate limit? I think it'd be a helpful debugging step for future problems.

**Comment by finestructure** [comment]
Ok, so this is what it looks like if it's a rate limiting error:

```
+ https://github.com/fourplusone/rasterize.git
ERROR: Rate Limit Exceeded /1000
🚨 https://github.com/fourplusone/rasterize.git: payload: {"message":"API rate limit exceeded for installation ID 7646295.","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}
The data couldn’t be read because it is missing.

1 package(s) out of 1 failed
Error: Process completed with exit code 1.
```

(from: https://github.com/SwiftPackageIndex/PackageList/pull/794/checks?check_run_id=1523644548)

So something else is wrong with these packages - and it seems to be either tied to CI or only happen at certain times.

**Comment by finestructure** [comment]
The error (or a similar error) is reproducible with PR https://github.com/SwiftPackageIndex/PackageList/pull/796

The action output https://github.com/SwiftPackageIndex/PackageList/pull/796/checks?check_run_id=1527515559 is

```
🚨 https://github.com/richardpiazza/LocaleSupport.git: dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: "The given data was not valid JSON.", underlyingError: Optional(Error Domain=NSCocoaErrorDomain Code=3840 "No value." UserInfo={NSDebugDescription=No value.})))
🚨 https://github.com/richardpiazza/VectorPlus.git: Request Timed Out
🚨 https://github.com/richardpiazza/XcodeServer.git: Request Timed Out
🚨 https://github.com/richardpiazza/SwiftColor.git: Request Timed Out
🚨 https://github.com/richardpiazza/SwiftSVG.git: Request Timed Out
🚨 https://github.com/richardpiazza/Swift2D.git: Request Timed Out

6 package(s) out of 7 failed
Error: Process completed with exit code 1.
```

whereas running locally shows:

```
*******************************************************************************
** IMPORTANT ******************************************************************

During validation, problem(s) were found:

* Found "https://github.com/autimatisering/IkigaJSON.git" but this redirected to "https://github.com/Joannis/IkigaJSON"
* Found "https://github.com/autimatisering/SMTPKitten.git" but this redirected to "https://github.com/Joannis/SMTPKitten"
* Found "https://github.com/adam-fowler/aws-cognito-authentication.git" but this redirected to "https://github.com/adam-fowler/soto-cognito-authentication"
* Found "https://github.com/iwasrobbed/Down.git" but this redirected to "https://github.com/johnxnguyen/Down"
* Found "https://github.com/adam-fowler/aws-cognito-authentication-kit.git" but this redirected to "https://github.com/adam-fowler/soto-cognito-authentication-kit"
* Found "https://github.com/autimatisering/VaporSMTPKit.git" but this redirected to "https://github.com/Joannis/VaporSMTPKit"
* The packages.json file was incorrectly sorted

These problems have been automatically fixed, but you'll need to commit the
changes that this script has made before creating the pull request.

Thanks!

*******************************************************************************
ERROR: 404 Not Found
🚨 https://github.com/richardpiazza/LocaleSupport.git: dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: "The given data was not valid JSON.", underlyingError: Optional(Error Domain=NSCocoaErrorDomain Code=3840 "No value." UserInfo={NSDebugDescription=No value.})))
🚨 https://github.com/richardpiazza/SwiftColor.git: dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: "The given data was not valid JSON.", underlyingError: Optional(Error Domain=NSCocoaErrorDomain Code=3840 "No value." UserInfo={NSDebugDescription=No value.})))
🚨 https://github.com/Mieraidihaimu/MMFoundation_swift.git: payload: {"message":"Not Found","documentation_url":"https://docs.github.com/rest/reference/repos#get-a-repository"}
The data couldn’t be read because it is missing.

3 package(s) out of 14 failed
```

`.validate-new.swift` I think is pointing to the underlying issue:

```
❯ swift .validate-new.swift 
Warning: Using anonymous authentication -- may run into rate limiting issues

ERROR: url not found (404): https://api.github.com/repos/Mieraidihaimu/MMFoundation_swift
```

The last nightly run fixed that issue and I suspect that it's been the cause for the multi-package validation to be failing. What's unclear is why the single package validation would be immune.

**Comment by finestructure** [comment]
After merging main into the branch both scripts pass:

```
❯ ./validate.swift
Warning: Using anonymous authentication -- may run into rate limiting issues

- https://github.com/richardpiazza/XCServerCoreData.git
- https://github.com/richardpiazza/XCServerAPI.git
- https://github.com/richardpiazza/SOSwiftVocabulary.git
- https://github.com/richardpiazza/BZipCompression.git
+ https://github.com/richardpiazza/CoordinatorPlus.git
+ https://github.com/richardpiazza/LocaleSupport.git
+ https://github.com/richardpiazza/SwiftColor.git
+ https://github.com/richardpiazza/SwiftSVG.git
+ https://github.com/richardpiazza/VectorPlus.git
+ https://github.com/richardpiazza/XcodeServer.git

6 package(s) passed
```

```
❯ swift .validate-new.swift
Warning: Using anonymous authentication -- may run into rate limiting issues

⚠️  Changes have been made to 'packages.json'. Your original version has been
⚠️  copied to 'package.backup.json'. Please commit the updated file.
ℹ️  Changes:
- ADDED   https://github.com/richardpiazza/CoordinatorPlus.git
- ADDED   https://github.com/richardpiazza/LocaleSupport.git
- ADDED   https://github.com/richardpiazza/SwiftColor.git
- ADDED   https://github.com/richardpiazza/SwiftSVG.git
- ADDED   https://github.com/richardpiazza/VectorPlus.git
- ADDED   https://github.com/richardpiazza/XcodeServer.git
- REMOVED https://github.com/richardpiazza/xcserverapi.git
- REMOVED https://github.com/richardpiazza/soswiftvocabulary.git
- REMOVED https://github.com/richardpiazza/xcservercoredata.git
- REMOVED https://github.com/richardpiazza/bzipcompression.git
✅ validation succeeded
```

Fixing that properly is going to be much easier in the new script, so it might be worth re-enabling the side-by-side and bringing it up to speed for switching over.

**Comment by finestructure** [comment]
There is something fishy going on. I just had two runs (one with each script) that passed, now following by this:

```
❯ env GITHUB_TOKEN=... swift validate.swift
- https://github.com/richardpiazza/XCServerCoreData.git
- https://github.com/richardpiazza/XCServerAPI.git
- https://github.com/richardpiazza/Swift2D
- https://github.com/richardpiazza/SOSwiftVocabulary.git
- https://github.com/richardpiazza/BZipCompression.git
+ https://github.com/richardpiazza/CoordinatorPlus.git
+ https://github.com/richardpiazza/LocaleSupport.git
+ https://github.com/richardpiazza/Swift2D.git
+ https://github.com/richardpiazza/SwiftColor.git
+ https://github.com/richardpiazza/SwiftSVG.git
+ https://github.com/richardpiazza/VectorPlus.git
+ https://github.com/richardpiazza/XcodeServer.git
🚨 https://github.com/richardpiazza/XcodeServer.git: dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: "The given data was not valid JSON.", underlyingError: Optional(Error Domain=NSCocoaErrorDomain Code=3840 "No value." UserInfo={NSDebugDescription=No value.})))

1 package(s) out of 7 failed
~/P/spi-package-list on  richardpiazza/main took 11s 
❯ env GITHUB_TOKEN=... swift .validate-new.swift 
dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: "The given data was not valid JSON.", underlyingError: Optional(Error Domain=NSCocoaErrorDomain Code=3840 "No value." UserInfo={NSDebugDescription=No value.})))
```

**Comment by finestructure** [comment]
Ok, we've got a working version now but I'd rather not talk about _how_ that's working... 😅

I'm fairly certain the issue is with the way we're making the `dataTask` synchronous. That's why the nightly has no issues with many more packages, it's all async and only waits at the end. I can't see anything wrong with the semaphore though, nor did using a DispatchGroup make any difference.

What I don't understand is why this would only happen in CI and not when running locally. CI is probably still on Catalina but that being the reason seems like a stretch. Will try this on Ceres.

For a fix: We could make the whole script async and wait only once - but that's quite the reshuffle. Or we leave the curl hackery in place but dear me if anyone sees that code...

What really bothers me about this script is that there are no tests and we can't add any as long as it's a script. (Not that tests would have helped with this particular issue but it's hard to touch this script and feel confident nothing else will break.)

So in summary:

1) leave it with curl (maybe adding a comment that we've heard of URLSession 😅)
2) rework it to be "more async"
3) reconsider making it a package that shares more/some/all code with the validator

1) Is easy but UGLY.

2) Is fiddly _and_ we're not even sure it'll fix the issue (although I suppose given the evidence it's quite likely).

3) Is quite a bit of work but has a maintenance upside.

We could also be quite confident it'll work, because we're doing the same thing in the nightly. In fact we could share that part and only have to worry about one validation.

The downside it that it'll add more clutter to the PackageList repository.

Shipping a binary would mitigate that. But looking at the validator binary, this will bloat the repo. Validator is 12MB (intel only - universal would probably close to double that but that wouldn't be necessary). However, it's probably going to run on a wider variety of OS versions (current and future) if we keep it a script or execute it with `swift run`.

**Comment by daveverwer** [comment]
> Ok, we've got a working version now but I'd rather not talk about _how_ that's working... 😅

The hourly commit messages made my day, as your sanity gradually departed and you came to terms with the situation 😂

> a) leave it with curl (maybe adding a comment that we've heard of URLSession 😅)

What I'd suggest is that we leave this solution there for a week or two, with a big comment explaining it either for people looking (which they won't, so this isn't anything to be concerned about) or for future-us. If it works, honestly? It doesn't matter much if it's ugly, this isn't the main app, it's a validation script.

If it doesn't work, which is a possibility when it's a hacky solution, then we can look at one of the other options in the future.

**Comment by finestructure** [comment]
Yeah, I'm leaning towards being pragmatic for now. And then look into making it a proper solution by folding it into the validator somehow.

I'll patch this up to it's mergeable.

**Comment by finestructure** [comment]
Well look at this. Seems like it _is_ a Catalina issue. Here's a run of `validate.swift` on Ceres:

```

Updated 0 paths from the index
- https://github.com/richardpiazza/XCServerCoreData.git
- https://github.com/richardpiazza/XCServerAPI.git
- https://github.com/richardpiazza/Swift2D
- https://github.com/richardpiazza/SOSwiftVocabulary.git
- https://github.com/richardpiazza/BZipCompression.git
- https://github.com/pusher/NWWebSocket
- https://github.com/kateinoigakukun/swift-indexstore
- https://github.com/glessard/CAtomics
- https://github.com/christopherweems/triple-q
- https://github.com/christopherweems/Resultto
- https://github.com/apple/swift-distributed-tracing-baggage
+ https://github.com/apple/swift-distributed-tracing-baggage.git
+ https://github.com/christopherweems/Resultto.git
+ https://github.com/christopherweems/triple-q.git
+ https://github.com/glessard/CAtomics.git
+ https://github.com/kateinoigakukun/swift-indexstore.git
+ https://github.com/pusher/NWWebSocket.git
+ https://github.com/richardpiazza/CoordinatorPlus.git
+ https://github.com/richardpiazza/LocaleSupport.git
+ https://github.com/richardpiazza/Swift2D.git
+ https://github.com/richardpiazza/SwiftColor.git
+ https://github.com/richardpiazza/SwiftSVG.git
+ https://github.com/richardpiazza/VectorPlus.git
+ https://github.com/richardpiazza/XcodeServer.git
🚨 https://github.com/glessard/CAtomics.git: Request Timed Out
🚨 https://github.com/kateinoigakukun/swift-indexstore.git: Request Timed Out
🚨 https://github.com/richardpiazza/XcodeServer.git: Request Timed Out
🚨 https://github.com/richardpiazza/LocaleSupport.git: Request Timed Out
🚨 https://github.com/richardpiazza/CoordinatorPlus.git: Request Timed Out

5 package(s) out of 13 failed
```

I'll see if we can use a Big Sur runner in CI.

**Comment by daveverwer** [comment]
How strange!

**Comment by finestructure** [comment]
Fixed by https://github.com/SwiftPackageIndex/PackageList/pull/798

**Comment by daveverwer** [comment]
"Fixed" 😬
